test/integration/dryrun/dryrun_test.go

```diff
@@ -0,0 +1,713 @@
+		skipUpdate: true,
+	},
+	// --
+
+}
+
+// Only add kinds to this list when this a virtual resource with get and create verbs that doesn't actually
+// store into it's kind.  We've used this downstream for mappings before.
+var kindWhiteList = sets.NewString()
+
+// namespace used for all tests, do not change this
+const testNamespace = "dryrunnamespace"
+
+func DryRunCreateTest(t *testing.T, rsc dynamic.ResourceInterface, obj *unstructured.Unstructured, gvResource schema.GroupVersionResource) {
+	if _, err := rsc.Create(obj, metav1.CreateOptions{DryRun: []string{metav1.DryRunAll}}); err != nil {
+		t.Fatalf("failed to dry-run create stub for %s: %#v", gvResource, err)
+	}
+
+	if _, err := rsc.Get(obj.GetName(), metav1.GetOptions{}); err == nil {
+		t.Fatalf("object shouldn't exist")
+	}
+}
+
+func DryRunPatchTest(t *testing.T, rsc dynamic.ResourceInterface, name string) {
+	patch := []byte(`{"metadata":{"annotations":{"patch": "true"}}}`)
+	obj, err := rsc.Patch(name, types.MergePatchType, patch, metav1.UpdateOptions{DryRun: []string{metav1.DryRunAll}})
+	if err != nil {
+		t.Fatalf("failed to dry-run patch object: %v", err)
+	}
+	if _, ok := obj.GetAnnotations()["patch"]; !ok {
+		t.Fatalf("dry-run patched annotations should be returned, got: %v", obj.GetAnnotations())
+	}
+	obj, err = rsc.Get(obj.GetName(), metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get object: %v", err)
+	}
+	if _, ok := obj.GetAnnotations()["patch"]; ok {
+		t.Fatalf("dry-run patched annotations should not be persisted, got: %v", obj.GetAnnotations())
+	}
+}
+
+func getReplicasOrFail(t *testing.T, obj *unstructured.Unstructured) int64 {
+	t.Helper()
+	replicas, found, err := unstructured.NestedInt64(obj.UnstructuredContent(), "spec", "replicas")
+	if err != nil {
+		t.Fatalf("failed to get int64 for replicas: %v", err)
+	}
+	if !found {
+		t.Fatal("object doesn't have spec.replicas")
+	}
+	return replicas
+}
+
+func setReplicasOrFail(t *testing.T, obj *unstructured.Unstructured, replicas int64) {
+	m, found, err := unstructured.NestedMap(obj.UnstructuredContent(), "spec")
+	if err != nil {
+		t.Fatalf("failed to get spec: %v", err)
+	}
+	if !found {
+		t.Fatal("object doesn't have spec")
+	}
+	m["replicas"] = replicas
+	unstructured.SetNestedMap(obj.UnstructuredContent(), m, "spec")
+}
+
+func DryRunScalePatchTest(t *testing.T, rsc dynamic.ResourceInterface, name string) {
+	obj, err := rsc.Get(name, metav1.GetOptions{}, "scale")
+	if err == nil {
+		replicas := getReplicasOrFail(t, obj)
+		patch := []byte(`{"spec":{"replicas":10}}`)
+		obj, err = rsc.Patch(name, types.MergePatchType, patch, metav1.UpdateOptions{DryRun: []string{metav1.DryRunAll}}, "scale")
```

Comment:
mind uniquely naming these objects instead of changing the meaning of obj as we go through?